jesd204: add basic finite-state-machine logic + linux bus register (part 2)

commodo opened the pull request:
This changeset integrates a few more bits of the framework.
Essentially here:
* the connections between devices are created from DT; at the time of the initial writing DT was used; conversion to more abstract DT/ACPI compat layer will be done later
* we try not to use the term link == connection; to avoid confusion between JESD204 links
   * a `connection` refers to the connection/link between 2 devices in terms of clock/inputs 
   * a `link` is a JESD204 link

The FSM should handle most of the logic/work-load of the framework when needing to configure/reconfigure links.

Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>

commodo commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -71,8 +83,120 @@ static struct jesd204_dev *jesd204_dev_find_by_of_node(struct device_node *np)
+		}
+		if (i != args->args_count)
+			continue;
+		return con;
```
i think i did it this way to make the patch more flexible when adding/removing conditions;
it's one of the patterns i use for search-loops;
then when adding a new match-condition, it shows as an +add-lines patch with no -remove lines;

so, now the last condition to succeed is `if (i == args->args_count)`, but later we could remove it, or add a new one;
not sure;

commodo commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -71,8 +83,120 @@ static struct jesd204_dev *jesd204_dev_find_by_of_node(struct device_node *np)
+	e->jdev = jdev;
+	list_add(&e->list, &con->dests);
+	con->dests_count++;
+
```
i need to check here; because `e->list` is also the list-head when starting to iterate the list;

commodo replied:
> i need to check here; because `e->list` is also the list-head when starting to iterate the list;

i think i got confused here with some older code;
this can be `entry`

commodo commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -88,6 +212,20 @@ static int jesd204_of_create_devices(void)
+		if (ret)
+			goto unlock;
+	}
+
```
`jesd204_of_create_devices` shouldn't be needto run from other places;
but that is still undefined;

commodo commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -105,6 +243,9 @@ struct jesd204_dev *jesd204_dev_register(struct device *dev,
 
+	if (!dev_is_jesd204_dev(dev))
+		return NULL;
+
```
no, not really;
some devices [JESD204 source-clocks] can/should operate without the framework;
so, for the AD9258 case: it can be a generic clock-source, or it can be used a JESD clock-source [SYSREF, lane-clk, etc];

but the driver will always call to register with the framework, and will just exit if no DT entry exists to put it into a JESD topology;
i'll admit, there is no good way [yet] to report if a device needs to be in a topology, and is not in the topology;
the burden of that requirement lies on DT;

nunojsa replied:
> i'll admit, there is no good way [yet] to report if a device needs to be in a topology, and is not in the topology;
the burden of that requirement lies on DT;

I was gonna to comment something similar. I would say that if a device is to be used as generic clock source or a JESD source looks like DT parameter. So, I guess we could move the logic of registering with framework [or not], for the drivers...

Either way, not to strong about this and leaving it as is also does not arm...

nunojsa replied:
Im a little bit confused with the commit message now that I saw it. Isn't it "don't register devices __without__ `jesd204-device` DT prop"? 

Or maybe we could put in another words: "Only register devices with jesd204-device` DT prop"...

commodo replied:
ack;
i wrote that description a while back;

commodo commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -135,6 +276,33 @@ static void jesd204_of_unregister_devices(void)
+		c = jdev->inputs[i];
+		list_for_each_entry_safe(e, e1, &c->dests, list) {
+			list_del(&e->list);
+			jesd204_dev_unregister(e->jdev);
```
so, yes;
recursion is messy;

but it also gives us the flexibility to define various types of topologies via DT;

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+#include "jesd204-priv.h"
+
+extern struct list_head jesd204_topologies;
+
```
i'll take a look;
this should only be shared by framework-core files;

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+	unsigned int i;
+	int ret = 0;
+
+	con = NULL;
```
i also was wondering [a few days ago] if i should leave it like this or move it;
i'll move it;

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+				     struct jesd204_fsm_data *data);
+
+const char *jesd204_state_str(enum jesd204_dev_state state)
+{
```
ack

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+			goto done;
+		ret = propagated_cb(con->owner, con, data);
+		if (ret)
+			goto done;
```
in this case, sure;

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+		if (c->jdev_top->nxt_state == JESD204_STATE_UNINIT)
+			return 0;
+		if (c->jdev_top->nxt_state != jdev_top->nxt_state)
+			return 0;
```
jdev_top should **not** be null;
if it is null, then it '''should''' also crash earlier in jesd204_fsm_cb()

nunojsa replied:
Ok, then I guess we can drop `if (jdev_top)...`

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+	if (c->jdev_top)
+		kref_put(&c->jdev_top->cb_ref,
+			 __jesd204_dev_top_fsm_change_cb);
+
```
> What will be the state of things if `jesd204_dev_validate_cur_state()` fails? I guess we would never reach 0? Is there anything to recover in that state?

this part is undefined yet;
currently, it's all-or-nothing ; but some polling-mechanism to recover may be an idea; or some userspace tool checks the status and re-triggers an entire FSM change; no idea;

nunojsa replied:
Hmm I see. I guess we should somehow notify the devices [in a topology] that we couldn't change state (not sure if it's done already).

commodo replied:
> Hmm I see. I guess we should somehow notify the devices [in a topology] that we couldn't change state (not sure if it's done already).

it isn't yet;
we should just be able to read the error code and determine which device errored;

commodo commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,353 @@
+			return ret;
+	} else
+		ret = JESD204_STATE_CHANGE_DONE;
+
```
ack

commodo commented:
> At this point this is what I can contribute :smile: . I would say that we reach the point where maybe it makes sense to have some doc (with some schemes) on how you see the framework. That could help the reviewing and also the code contributions....

I'll need to search for some old thoughts/ideas.
If not, then re-draw some.

So, at this point, this is just some generic graph-mechanism, that will eventually have some hooks defined.

commodo commented:
Changelog v1 -> v2:
* moved `con = NULL` init to declaration
* removed `if (jdev_top)` check
* changed `goto done` -> `break`
* renamed `list` -> `entry` field
* addressed checkpatch complaint about curly braces;
* added getter for `jesd204_topologies`

nunojsa commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -0,0 +1,349 @@
+		if (c->jdev_top->nxt_state == JESD204_STATE_UNINIT)
+			return 0;
+		if (c->jdev_top->nxt_state != jdev_top->nxt_state)
+			return 0;
```
Hmm. Couldn't we do this a bit simpler? Now that I look at it, i guess that we could ditch the second `if (c->jdev_top->nxt_state != jdev_top->nxt_state)`. We are already doing `if (c->jdev_top != jdev_top)`. So if the objects are the same...

commodo replied:
ack

nunojsa commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -71,8 +83,120 @@ static struct jesd204_dev *jesd204_dev_find_by_of_node(struct device_node *np)
+		if (ret) {
+			of_node_put(args.np);
+			return ret;
+		}
```
Am I missing something? I really think we should also call `of_node_put(args.np);` when moving into the next iteration...

commodo replied:
oops;
forgot this one;
sorry;

commodo commented:
Changelog v2 -> v3:
* added missing `of_node_put(args.np)`
* removed superfluous if check
* updated commit desc `jesd204: register only devices with `jesd204-device` DT prop`

commodo commented:
thanks for the review

let's see how this goes;
